diff --git a/cutils.c b/cutils.c
index a78e36f..f02d95e 100644
--- a/cutils.c
+++ b/cutils.c
@@ -30,7 +30,7 @@
 #include "cutils.h"
 
 #ifdef _MSC_VER
-
+#ifdef USE_GETTIMEOFDAY_INTERNAL
  // From: https://stackoverflow.com/a/26085827
 int gettimeofday(struct timeval * tp, struct timezone * tzp)
 {
@@ -51,6 +51,7 @@ int gettimeofday(struct timeval * tp, struct timezone * tzp)
   return 0;
 }
 #endif
+#endif
 
 
 void pstrcpy(char *buf, int buf_size, const char *str)
diff --git a/cutils.h b/cutils.h
index 51f0be8..f6ecb7e 100644
--- a/cutils.h
+++ b/cutils.h
@@ -29,14 +29,34 @@
 #include <inttypes.h>
 #include <string.h>
 
+#if defined(_WIN32)
+    #include <windows.h>
+
+    #if __has_include(<intrin.h>)
+        #include <intrin.h>
+    #else
+        #if defined(__GNUC__)
+            #include <cpuid.h>
+        #endif
+    #endif
+#endif
+
 #ifdef _MSC_VER
-  #include <windows.h>
-  #include <intrin.h>
-  #ifndef alloca
-    #define alloca(s) _alloca(s)
-  #endif
-#else 
-  #include <sys/time.h>
+    #ifndef alloca
+        #define alloca(s) _alloca(s)
+    #endif
+
+    #include <BaseTsd.h>
+    typedef SSIZE_T ssize_t;
+
+    #if __has_include(<sys/time.h>)
+        #include <sys/time.h>
+    #else
+        #define USE_GETTIMEOFDAY_INTERNAL
+        int gettimeofday(struct timeval * tp, struct timezone * tzp);
+    #endif
+#else
+    #include <sys/time.h>
 #endif
 
 /* set if CPU is big endian */
@@ -50,7 +70,9 @@
   #define __maybe_unused
   #define __attribute__(x)
   #define __attribute(x)
-  typedef intptr_t ssize_t;
+  #ifndef _MSC_VER
+    typedef intptr_t ssize_t;
+  #endif
   #define printf_like(A, B) /*__attribute__((format(printf, (A), (B))))*/
 #else
   #define likely(x)       __builtin_expect(!!(x), 1)
diff --git a/quickjs-libc.c b/quickjs-libc.c
index a6e8f36..293a152 100644
--- a/quickjs-libc.c
+++ b/quickjs-libc.c
@@ -34,37 +34,55 @@
 #include <signal.h>
 #include <limits.h>
 #include <sys/stat.h>
+
 #if defined(_WIN32)
-  #include <windows.h>
-  #include <conio.h>
-  #include <io.h>
-  #include <fcntl.h>
-  #include <sys/types.h>
-  #include <sys/stat.h>
-  #include <sys/utime.h>
-  #include "win/dirent.h"
-  #ifndef PATH_MAX
-    #define PATH_MAX MAX_PATH
-  #endif
-  #define popen _popen
-  #define pclose _pclose
+    #include <windows.h>
+    #include <conio.h>
+    #include <io.h>
+    #include <fcntl.h>
+    #include <sys/types.h>
+    #include <sys/stat.h>
+    #include <sys/utime.h>
+
+    #if __has_include(<dirent.h>)
+        #include <dirent.h>
+    #else
+        #include "win/dirent.h"
+    #endif
+
+    #if __has_include(<unistd.h>)
+        #include <unistd.h>
+    #endif
+
+    #ifndef PATH_MAX
+        #define PATH_MAX MAX_PATH
+    #endif
+
+    #define popen _popen
+    #define pclose _pclose
 #else
-  #include <dirent.h>
-  #include <unistd.h>
-  #include <sys/time.h>
-#include <dlfcn.h>
-#include <termios.h>
-#include <sys/ioctl.h>
-#include <sys/wait.h>
-
-#if defined(__APPLE__)
-typedef sig_t sighandler_t;
-#if !defined(environ)
-#include <crt_externs.h>
-#define environ (*_NSGetEnviron())
-#endif
-#endif /* __APPLE__ */
-
+    #include <dirent.h>
+    #include <unistd.h>
+    #include <sys/time.h>
+    #include <dlfcn.h>
+    #include <termios.h>
+    #include <sys/ioctl.h>
+    #include <sys/wait.h>
+
+    #if defined(__APPLE__)
+        typedef sig_t sighandler_t;
+        #if !defined(environ)
+            #include <crt_externs.h>
+            #define environ (*_NSGetEnviron())
+        #endif
+    #endif /* __APPLE__ */
+
+    #if defined(__FreeBSD__)
+        typedef sig_t sighandler_t;
+        __BEGIN_DECLS
+        extern char **environ;
+        __END_DECLS
+    #endif  /* __FreeBSD__ */
 #endif
 
 #if !defined(_WIN32)
@@ -142,6 +160,18 @@ typedef struct JSThreadState {
 static uint64_t os_pending_signals;
 static int (*os_poll_func)(JSContext *ctx);
 
+#if defined(_WIN32)
+static char *realpath(const char *path, char *buf)
+{
+    if (!_fullpath(buf, path, PATH_MAX)) {
+        errno = ENOENT;
+        return NULL;
+    } else {
+        return buf;
+    }
+}
+#endif
+
 static void js_std_dbuf_init(JSContext *ctx, DynBuf *s)
 {
     dbuf_init2(s, JS_GetRuntime(ctx), (DynBufReallocFunc *)js_realloc_rt);
@@ -540,7 +570,7 @@ int js_module_set_import_meta(JSContext *ctx, JSValueConst func_val,
         return -1;
     if (!strchr(module_name, ':')) {
         strcpy(buf, "file://");
-#if !defined(_WIN32)
+// #if !defined(_WIN32)
         /* realpath() cannot be used with modules compiled with qjsc
            because the corresponding module source code is not
            necessarily present */
@@ -552,7 +582,7 @@ int js_module_set_import_meta(JSContext *ctx, JSValueConst func_val,
                 return -1;
             }
         } else
-#endif
+// #endif
         {
             pstrcat(buf, sizeof(buf), module_name);
         }
@@ -2658,18 +2688,6 @@ static JSValue js_os_sleep(JSContext *ctx, JSValueConst this_val,
     return JS_NewInt32(ctx, ret);
 }
 
-#if defined(_WIN32)
-static char *realpath(const char *path, char *buf)
-{
-    if (!_fullpath(buf, path, PATH_MAX)) {
-        errno = ENOENT;
-        return NULL;
-    } else {
-        return buf;
-    }
-}
-#endif
-
 /* return [path, errorcode] */
 static JSValue js_os_realpath(JSContext *ctx, JSValueConst this_val,
                               int argc, JSValueConst *argv)
diff --git a/quickjs.c b/quickjs.c
index eb6642b..b19f30b 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -285,6 +285,9 @@ struct JSRuntime {
     JSHostPromiseRejectionTracker *host_promise_rejection_tracker;
     void *host_promise_rejection_tracker_opaque;
 
+    JSHostPromiseRejectionTracker *host_unhandled_promise_rejection_tracker;
+    void *host_unhandled_promise_rejection_tracker_opaque;
+
     struct list_head job_list; /* list of JSJobEntry.link */
 
     JSModuleNormalizeFunc *module_normalize_func;
@@ -7979,6 +7982,8 @@ static JSValue JS_GetPropertyValue(JSContext *ctx, JSValueConst this_obj,
         uint32_t idx, len;
         /* fast path for array access */
         p = JS_VALUE_GET_OBJ(this_obj);
+        if (unlikely(!p->fast_array))
+            goto slow_path;
         idx = JS_VALUE_GET_INT(prop);
         len = (uint32_t)p->u.array.count;
         if (unlikely(idx >= len))
@@ -42379,6 +42384,14 @@ static JSValue js_math_random(JSContext *ctx, JSValueConst this_val,
 static double js_math_floor(double x) { return floor(x); }
 static double js_math_ceil(double x) { return ceil(x); }
 
+#if defined(_MSC_VER)
+#ifdef log2
+#pragma push_macro("log2")
+#undef log2
+#define MSC_LOG2_UNDEFINED
+#endif
+#endif
+
 static const JSCFunctionListEntry js_math_funcs[] = {
     JS_CFUNC_MAGIC_DEF("min", 2, js_math_min_max, 0 ),
     JS_CFUNC_MAGIC_DEF("max", 2, js_math_min_max, 1 ),
@@ -42428,6 +42441,12 @@ static const JSCFunctionListEntry js_math_funcs[] = {
     JS_PROP_DOUBLE_DEF("SQRT2", 1.4142135623730951, 0 ),
 };
 
+#if defined(_MSC_VER)
+#ifdef MSC_LOG2_UNDEFINED
+#pragma pop_macro("log2")
+#endif
+#endif
+
 static const JSCFunctionListEntry js_math_obj[] = {
     JS_OBJECT_DEF("Math", js_math_funcs, countof(js_math_funcs), JS_PROP_WRITABLE | JS_PROP_CONFIGURABLE ),
 };
@@ -46739,6 +46758,7 @@ typedef struct JSPromiseData {
     struct list_head promise_reactions[2];
     BOOL is_handled; /* Note: only useful to debug */
     JSValue promise_result;
+    JSContext * ctx;
 } JSPromiseData;
 
 typedef struct JSPromiseFunctionDataResolved {
@@ -46819,6 +46839,14 @@ void JS_SetHostPromiseRejectionTracker(JSRuntime *rt,
     rt->host_promise_rejection_tracker_opaque = opaque;
 }
 
+void JS_SetHostUnhandledPromiseRejectionTracker(JSRuntime *rt,
+                                       JSHostPromiseRejectionTracker *cb,
+                                       void *opaque)
+{
+    rt->host_unhandled_promise_rejection_tracker = cb;
+    rt->host_unhandled_promise_rejection_tracker_opaque = opaque;
+}
+
 static void fulfill_or_reject_promise(JSContext *ctx, JSValueConst promise,
                                       JSValueConst value, BOOL is_reject)
 {
@@ -46829,6 +46857,7 @@ static void fulfill_or_reject_promise(JSContext *ctx, JSValueConst promise,
 
     if (!s || s->promise_state != JS_PROMISE_PENDING)
         return; /* should never happen */
+
     set_value(ctx, &s->promise_result, JS_DupValue(ctx, value));
     s->promise_state = JS_PROMISE_FULFILLED + is_reject;
 #ifdef DUMP_PROMISE
@@ -47023,6 +47052,14 @@ static void js_promise_finalizer(JSRuntime *rt, JSValue val)
 
     if (!s)
         return;
+
+    if (s->promise_state == JS_PROMISE_REJECTED && !s->is_handled) {
+        if (rt->host_unhandled_promise_rejection_tracker) {
+            rt->host_unhandled_promise_rejection_tracker(s->ctx, val, s->promise_result, FALSE,
+                                                         rt->host_unhandled_promise_rejection_tracker_opaque);
+        }
+    }
+
     for(i = 0; i < 2; i++) {
         list_for_each_safe(el, el1, &s->promise_reactions[i]) {
             JSPromiseReactionData *rd =
@@ -47073,6 +47110,7 @@ static JSValue js_promise_constructor(JSContext *ctx, JSValueConst new_target,
     s = js_mallocz(ctx, sizeof(*s));
     if (!s)
         goto fail;
+    s->ctx = ctx;
     s->promise_state = JS_PROMISE_PENDING;
     s->is_handled = FALSE;
     for(i = 0; i < 2; i++)
diff --git a/quickjs.h b/quickjs.h
index 1193a88..6a59807 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -1005,6 +1005,7 @@ typedef void JSHostPromiseRejectionTracker(JSContext *ctx, JSValueConst promise,
                                            JSValueConst reason,
                                            JS_BOOL is_handled, void *opaque);
 void JS_SetHostPromiseRejectionTracker(JSRuntime *rt, JSHostPromiseRejectionTracker *cb, void *opaque);
+void JS_SetHostUnhandledPromiseRejectionTracker(JSRuntime *rt, JSHostPromiseRejectionTracker *cb, void *opaque);
 
 /* return != 0 if the JS code needs to be interrupted */
 typedef int JSInterruptHandler(JSRuntime *rt, void *opaque);
